### ISSUE TITLE
[Feature]Add definition revision to support reasonable definition iteration

### ISSUE THREAD

**Issue by ZhongsJie** [kubevela/kubevela#3573]
**Is your feature request related to a problem? Please describe.**

At present, if the native definition of kubevela needs to be modified, such as `{} -> []`, the users who use the former version will be affected. Similar modifications to the definition will introduce break change.

In the long-term, to ensure compatibility, redefining a field to meet special scenarios may result in complex structures of definition.

I hope that kubevela can provide a definition version management mechanism. It's necessary to explore ways to ensure compatibility. At the same time, it has reasonable scalability to meet the needs of differentiated scenarios. IMO, two core points need to be guaranteed:

1. The definition revision of the previous users should be limited.
2. Later users can custom definition revision to meet specific scenarios.

**Describe the solution you'd like**

Use `groupVersion` to standardize template rendering process, which can be redefined after major changes. At the same time, there will be `preferredVersion` for different definitions, which can be overridden by customization. Users are not restricted from using definitions belonging to diverse groups.

**Describe alternatives you've considered**

Define the version of definitions through label or other forms, and set the `groupVersion` at the top level to constrain the rendering process. 

*Priority*: The version of the custom-definition is higher than the `preferredVersion` that exists by default.

**Additional context**

**Comment by FogDong** [comment]
Thanks for the proposal!

So, all of our definitions now should be set to `v1 groupVersion`.
And I think for every definition, we need to set a `stable` or `core` version, and for those applications without versions specified, we use a `stable version`.

For example, there is an application without a definition version:

```yaml
components:
  - name: test
     type: webservice
```

If the stable version of `webservice` is `v1`, then we use `webservice@v1`.

For a new application, which definition has two group versions, `corev1` and `v2alpha`. if we want to use definitions in `v2alpha`, we need to specify it in component type like:

```yaml
components:
  - name: test
     type: webservice@v2alpha1
```

But, there's still some gap if there's some break change between `webservice@v1` and `webservice@v2alpha2` and we want to upgrade `webservice@v2alpah2` to a stable version.

In that way, we may need a `converter` to convert `v1` to `v2alpha2`.

For examples, if the field is upgrade from `string` to `[]string`, we need to convert it in the `converter`.

**Comment by ZhongsJie96** [comment]
> Thanks for the proposal!
> 
> So, all of our definitions now should be set to `v1 groupVersion`. And I think for every definition, we need to set a `stable` or `core` version, and for those applications without versions specified, we use a `stable version`.
> 
> For example, there is an application without a definition version:
> 
> ```yaml
> components:
>   - name: test
>      type: webservice
> ```
> 
> If the stable version of `webservice` is `v1`, then we use `webservice@v1`.
> 
> For a new application, which definition has two group versions, `corev1` and `v2alpha1`. if we want to use definitions in `v2alpha1`, we need to specify it in component type like:
> 
> ```yaml
> components:
>   - name: test
>      type: webservice@v2alpha1
> ```
> 
> But, there's still some gap if there's some breaking change between `webservice@v1` and `webservice@v2alpha1` and we want to upgrade `webservice@v2alpah2` to a stable version.
> 
> In that way, we may need a `converter` to convert `v1` to `v2alpha1`.
> 
> For example, if the field is upgraded from `string` to `[]string`, we need to convert it in the `converter`.

IMO, this is a relatively good implementation for the definition version. The conversion between versions can be achieved by adding conversion logic to the new version template, the corresponding conversion processing is performed by comparing the version number when the user modifies the version. However, this implementation will increase the complexity of the template, and it is necessary to consider the conversion problem of multiple previous versions

**Comment by yangsoon** [comment]
In fact, kubevela has implemented basic definition revision features, you can specify to create a specific version of the definition and specify it in the Application through the `@` .

https://kubevela.io/docs/platform-engineers/x-def-version
https://github.com/oam-dev/kubevela/pull/2044

**Comment by wonderflow** [comment]
Yes, just like @yangsoon said, we already have a version mechanism for x-definition.

I think just add a mutating webhook can make the things better.


![image](https://user-images.githubusercontent.com/2173670/162677641-8b439ede-6d29-4af5-9293-77e49c1040d6.png)

**Comment by wonderflow** [comment]
ref https://github.com/oam-dev/kubevela/issues/2620

**Comment by wonderflow** [comment]
We need a more concrete design before Implementation.

**Comment by wonderflow** [comment]
As @morigs  said in https://github.com/oam-dev/spec/issues/483 , we should fix the definition revision issue.

Just like the github action versioning feature:

- Definition declared with no revision:  `type: my-def`, it should always use the latest definition version.
- Definition declared with revision but only Major version: `type: my-def@v1`, it will automatically update the revision when definition changed from v1.0 to v1.1 , or from v1.0.1 to v1.0.2. The same works when specifying with `type: my-def@v1.0`, it can automaticaly update from v1.0.1 to v1.0.2 while it don't when v1.1 created.

We also need to support specify version for definition in the following ways.

1. allow define version in definitions

task: {
	type: "component"
        version: "v1.0.0"
	annotations: {}
	labels: {}
	description: "Describes jobs that run code or a script to completion."
}
template: {
	parameter: {
	}
}

2. allow specify version when update definition in CLI

```
vela def apply --version v1.0.1  mydef.cue
```

It should report an override promt when the definition already has a version.


3. Add validation/mutating check on definition webhook that only increase version is allowed, definitions without version should be marked as `v0.1.0` automatically, any update without will increase the patch version automatically.

**Comment by morigs** [comment]
@wonderflow I wanted to decompose the task, but tbh I feel overwhelmed...
Retrieving version from file (`pkg/definition/definition.go`) and CLI (`references/cli/def.go`) looks fairly simple. I only have one question: where should be revision stored in `Definition` struct? Should it be stored in `oam.AnnotationDefinitionRevisionName` annotation?

However, I struggle to understand which parts of runtime should be updated for semver revision support...
I have figured out that definition controllers are in `pkg/controller/core.oam.dev/v1alpha2/core/components/...`, and that they update revisions using `pkg/controller/core.oam.dev/v1alpha2/core/revison.go`. These functions in turn use `Revision` struct from `apis/core.oam.dev/common/types.go`. And this struct is used everywhere... So I feel like changes to it will affect nit only definitions, but also other resources.

Also, I wasn't able to locate logic responsible for choosing a specific revision based on string in `type`.

Could you please give me guidance?

**Comment by morigs** [comment]
@wonderflow Hi, I see that the issue is in the 1.9 milestone. Is it in progress. Do you need any help?

**Comment by Somefive** [comment]
@morigs Sorry for the late response. KubeVela recently is facing shortage of hands.

I think by far we might not necessary to add too much things to the Definition struct now. We have XXXDefinition (like ComponentDefinition) and DefinitionRevision now. The revision of XXXDefinition is managed by the DefinitionRevision object. So currently the below machenisim is working.

1. First create a ComponentDefinition "kvcm".
```cue
"kvcm": {
	type: "component"
	annotations: {}
	labels: {}
	description: "Describes"
	attributes: workload: type: "autodetects.core.oam.dev"
}
template: {
	output: {
		apiVersion: "v1"
		kind: "ConfigMap"
		data: key: "value1"
	}
	parameter: {}
}
```
2. Update it.
```cue
"kvcm": {
	type: "component"
	annotations: {}
	labels: {}
	description: "Describes"
	attributes: workload: type: "autodetects.core.oam.dev"
}
template: {
	output: {
		apiVersion: "v1"
		kind: "ConfigMap"
		data: key: "value2"
	}
	parameter: {}
}
```
3. Update it again.
```cue
"kvcm": {
	type: "component"
	annotations: {}
	labels: {}
	description: "Describes"
	attributes: workload: type: "autodetects.core.oam.dev"
}
template: {
	output: {
		apiVersion: "v1"
		kind: "ConfigMap"
		data: key: "value3"
	}
	parameter: {}
}
```
4. You can see there is three DefinitionRevision created.
```bash
$ k get definitionrevision -n vela-system | grep kvcm                                                                  (okg-1)
kvcm-v1                            1          e208605f56c7241e   Component
kvcm-v2                            2          8772fe381599ef8d   Component
kvcm-v3                            3          b4cda672a8e57d35   Component
```
6. You can create an application using the latest "value3" by
```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: kvcm
spec:
  components:
    - type: kvcm
      name: kvcm
```
Or using the previous version
```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: kvcm
spec:
  components:
    - type: kvcm@v2
      name: kvcm
```

The thing @wonderflow mentions that we should improve is to let user be able to specify the version number as an alternative choice for the auto-generated one. So we can break down this problem as follows:

1. Add a new flag "--revision" to the "vela def apply" command. With this flag, the XXXDefinition will not be updated/created. Instead, it will create a DefinitionRevision. The manually specified revision name should be placed on the DefinitionRevision's label `definition.oam.dev/revision-name`.
2. Support the "--revision" flag for "vela def get". Add "vela def revisions list [DEFNAME] [-t/--type component/trait/policy/workflow-step]", "vela def revisions get [DEFNAME@REVNAME]" (this should work as an alias of "vela def get --revision"), "vela def revisions del [DEFNAME@REVNAME]" and "vela def revisions apply" (this should work as an alias of "vela def apply --revision").
3. Modify the function "fetchDefinitionRev" in "pkg/oam/util/helper.go" and support loading definition with the label selector `definition.oam.dev/revision-name`.

**Comment by anoop2811** [comment]
I believe this is fixed by https://github.com/kubevela/kubevela/issues/6435. Please feel to reopen if it doesnt with more commentary
